README
# TSL2591 Python Library

[![GitHub stars](https://img.shields.io/github/stars/maxlklaxl/python-tsl2591.svg)](https://github.com/maxlklaxl/python-tsl2591/stargazers)
[![GitHub forks](https://img.shields.io/github/forks/maxlklaxl/python-tsl2591.svg)](https://github.com/maxlklaxl/python-tsl2591/network)
[![GitHub issues](https://img.shields.io/github/issues/maxlklaxl/python-tsl2591.svg)](https://github.com/maxlklaxl/python-tsl2591/issues)
[![Inspired by J0s3f](https://img.shields.io/badge/Inspired%20by-adafruit-blue.svg)](https://gitgud.io/adafruit/Adafruit_TSL2591_Library)

This is a simple python library for the Adafruit TSL2591 breakout board based on the [Arduino library](https://github.com/adafruit/Adafruit_TSL2591_Library) from Adafruit. It was developed to work on a Raspberry PI.

## Installation

### Step 1: Enable I2C

You can enable I2C on the Raspberry Pi by following the instructions on [Adafruit](https://learn.adafruit.com/adafruits-raspberry-pi-lesson-4-gpio-setup/configuring-i2c).

The quick version is:

1. Run `sudo raspi-config`
2. Select `Advanced Options`
3. Enable I2C
4. Reboot (`sudo reboot`)

When testing I2C (`sudo i2cdetect -y 1`), you should see at least one connected device, your TSL2591 at `0x29`. For more information, see the [FAQ](#i2c-check-for-static-address).

### Step 2: Install System dependencies

Prior to using this library, you will need the following packages installed on your Raspberry Pi.

```bash
sudo apt-get install python3-dev python3-smbus libffi-dev libssl-dev
```

### Step 3: Install Python dependencies

To install the Python module, download this repository and run:

```
python setup.py install
```

## Quickstart

```
import tsl2591

tsl = tsl2591.Tsl2591()  # initialize
full, ir = tsl.get_full_luminosity()  # read raw values (full spectrum and ir spectrum)
lux = tsl.calculate_lux(full, ir)  # convert raw values to lux
print lux, full, ir
```

## FAQ

### Fatal error

If you do not have those Raspbian packages installed prior to installing this library, you will run into errors that look similar to this.

```
fatal error: ffi.h: No such file or directory
     #include <ffi.h>
                     ^
compilation terminated.
error: command 'arm-linux-gnueabihf-gcc' failed with exit status 1
```

### I2C Check for Static Address

Because the TSL2591 connects via I2C, it's always good to run the I2C detection to verify the address is being read. Unlike the TSL2561 with programmable addresses, the TSL2591's address is hard coded and cannot be changed - thus it will always show `0x29`.

In the example output below, you can see there are two I2C devices detected, one being the TSL2591.

```
pi@raspberrypi:~ $ sudo i2cdetect -y 1
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:          -- -- -- -- -- -- -- -- -- -- -- -- --
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
20: -- -- -- -- -- -- -- -- -- 29 -- -- -- -- -- --
30: -- -- -- -- -- -- -- -- -- -- -- -- 3c -- -- --
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
70: -- -- -- -- -- -- -- --
```

## License

Python files in this repository are released under the [MIT license](LICENSE.md).


CHANGELOG.md
<a name="0.2.0"></a>
## [0.2.0] - 2019-06-05
- Updated README for proper pip instructions and sample
- Deprecated smbus requirements (noted in FAQ)
- Fix version import

<a name="0.1.5"></a>
## [0.1.5] - 2019-06-05
- [Updating name, adding package scaffolding from CookieCutter](https://github.com/audreyr/cookiecutter-pypackage)
- Added Travis CI config (removed a test because it only works on Raspbian or devices with I2C enabled)
- Updated scaffold from existing files
- Updating package name for consistency (python_tsl2591)
- Published to [PyPi](https://pypi.org/project/python-tsl2591/)

<a name="0.1.4"></a>
## [0.1.4] - 2019-06-04
- Added CHANGELOG
- Published to [PyPi](https://pypi.org/project/tsl2591/)

<a name="0.1.3"></a>
## [0.1.3] - 2019-06-04
- Added packaging files for PyPi
- Added Manifest, updated README
- Updated setup.py based on https://github.com/kennethreitz/setup.py (modernized)
- Replace smbus/smbus-cffi with smbus2, a Python-only derivative. Removes the requirement to have smbus package installed on system.
- Linted for PEP8
- Updated class name for clarity, clean up variables
- Added instructions for installation
- Added common get_current() function with JSON response option
- Fixes integration_time plus **1** second blocked reliable measures (issue #10)

<a name="0.0.1"></a>
## [0.0.1] - 2016-03-06
- [Changed order of reading the registers in get_full_luminosity(self)](https://github.com/maxlklaxl/python-tsl2591/pull/6)
- [Fix get_luminostiy function and self test code](https://github.com/maxlklaxl/python-tsl2591/pull/4)
- [Corrects an error in the get_gain code](https://github.com/maxlklaxl/python-tsl2591/pull/3)

<a name="0.0.0"></a>
## 0.0.0 - 2015-05-07
- Initial release

tsl2591/main.py
'''
This code is an adaptation of the Arduino_TSL2591 library from
adafruit: https://github.com/adafruit/Adafruit_TSL2591_Library

For configuring I2C on Raspberry Pi
https://learn.adafruit.com/adafruits-raspberry-pi-lesson-4-gpio-setup/configuring-i2c

Datasheet
https://learn.adafruit.com/adafruit-tsl2591/downloads
'''
from __future__ import print_function

import time
import json
import smbus

# *************************************************
# ******* MACHINE VARIABLES (DO NOT TOUCH) ********
# *************************************************
VISIBLE = 2  # channel 0 - channel 1
INFRARED = 1  # channel 1
FULLSPECTRUM = 0  # channel 0

ADDR = 0x29
READBIT = 0x01
COMMAND_BIT = 0xA0  # bits 7 and 5 for 'command normal'
CLEAR_BIT = 0x40  # Clears any pending interrupt (write 1 to clear)
WORD_BIT = 0x20  # 1 = read/write word (rather than byte)
BLOCK_BIT = 0x10  # 1 = using block read/write
ENABLE_POWERON = 0x01
ENABLE_POWEROFF = 0x00
ENABLE_AEN = 0x02
ENABLE_AIEN = 0x10
CONTROL_RESET = 0x80
LUX_DF = 408.0
LUX_COEFB = 1.64  # CH0 coefficient
LUX_COEFC = 0.59  # CH1 coefficient A
LUX_COEFD = 0.86  # CH2 coefficient B

REGISTER_ENABLE = 0x00
REGISTER_CONTROL = 0x01
REGISTER_THRESHHOLDL_LOW = 0x02
REGISTER_THRESHHOLDL_HIGH = 0x03
REGISTER_THRESHHOLDH_LOW = 0x04
REGISTER_THRESHHOLDH_HIGH = 0x05
REGISTER_INTERRUPT = 0x06
REGISTER_CRC = 0x08
REGISTER_ID = 0x0A
REGISTER_CHAN0_LOW = 0x14
REGISTER_CHAN0_HIGH = 0x15
REGISTER_CHAN1_LOW = 0x16
REGISTER_CHAN1_HIGH = 0x17
# *****************************************
# ******* END OF MACHINE VARIABLES ********
# *****************************************

# Integration time
# The integration time can be set between 100 and 600ms,
# and the longer the integration time the more light the
# sensor is able to integrate, making it more sensitive in
# low light the longer the integration time.
INTEGRATIONTIME_100MS = 0x00
INTEGRATIONTIME_200MS = 0x01
INTEGRATIONTIME_300MS = 0x02
INTEGRATIONTIME_400MS = 0x03
INTEGRATIONTIME_500MS = 0x04
INTEGRATIONTIME_600MS = 0x05

# Gain
# The gain can be set to one of the following values
# (though the last value, MAX, has limited use in the
# real world given the extreme amount of gain applied):
# GAIN_LOW: Sets the gain to 1x (bright light)
# GAIN_MEDIUM: Sets the gain to 25x (general purpose)
# GAIN_HIGH: Sets the gain to 428x (low light)
# GAIN_MAX: Sets the gain to 9876x (extremely low light)
GAIN_LOW = 0x00
GAIN_MED = 0x10
GAIN_HIGH = 0x20
GAIN_MAX = 0x30


class Tsl2591(object):
    '''
    An object class containing a series of methods to enable easy
    interaction with the sensor.
    '''
    def __init__(
            self,
            i2c_bus=1,
            sensor_address=0x29,
            integration=INTEGRATIONTIME_200MS,
            gain=GAIN_MED
    ):
        self.bus = smbus.SMBus(i2c_bus)
        self.sender_address = sensor_address
        self.integration_time = integration
        self.gain = gain
        self.set_timing(self.integration_time)
        self.set_gain(self.gain)
        self.disable()  # to be sure

    def set_timing(self, integration):
        self.enable()
        self.integration_time = integration
        self.bus.write_byte_data(
            self.sender_address,
            COMMAND_BIT | REGISTER_CONTROL,
            self.integration_time | self.gain
        )
        self.disable()

    def get_timing(self):
        return self.integration_time

    def set_gain(self, gain):
        self.enable()
        self.gain = gain
        self.bus.write_byte_data(
            self.sender_address,
            COMMAND_BIT | REGISTER_CONTROL,
            self.integration_time | self.gain
        )
        self.disable()

    def get_gain(self):
        return self.gain

    def calculate_lux(self, full, ir):
        # Check for overflow conditions first
        if (full == 0xFFFF) | (ir == 0xFFFF):
            return 0

        case_integ = {
            INTEGRATIONTIME_100MS: 100.,
            INTEGRATIONTIME_200MS: 200.,
            INTEGRATIONTIME_300MS: 300.,
            INTEGRATIONTIME_400MS: 400.,
            INTEGRATIONTIME_500MS: 500.,
            INTEGRATIONTIME_600MS: 600.,
        }
        if self.integration_time in case_integ.keys():
            atime = case_integ[self.integration_time]
        else:
            atime = 100.

        case_gain = {
            GAIN_LOW: 1.,
            GAIN_MED: 25.,
            GAIN_HIGH: 428.,
            GAIN_MAX: 9876.,
        }

        if self.gain in case_gain.keys():
            again = case_gain[self.gain]
        else:
            again = 1.

        # cpl = (ATIME * AGAIN) / DF
        cpl = (atime * again) / LUX_DF
        lux1 = (full - (LUX_COEFB * ir)) / cpl

        lux2 = ((LUX_COEFC * full) - (LUX_COEFD * ir)) / cpl

        # The highest value is the approximate lux equivalent
        return max([lux1, lux2])

    def enable(self):
        self.bus.write_byte_data(
            self.sender_address,
            COMMAND_BIT | REGISTER_ENABLE,
            ENABLE_POWERON | ENABLE_AEN | ENABLE_AIEN
        )  # Enable

    def disable(self):
        self.bus.write_byte_data(
            self.sender_address,
            COMMAND_BIT | REGISTER_ENABLE,
            ENABLE_POWEROFF
        )

    def get_full_luminosity(self):
        self.enable()
        # not sure if we need it "// Wait x ms for ADC to complete"
        time.sleep(0.105+0.100*self.integration_time)
        full = self.bus.read_word_data(
            self.sender_address, COMMAND_BIT | REGISTER_CHAN0_LOW
        )
        ir = self.bus.read_word_data(
            self.sender_address, COMMAND_BIT | REGISTER_CHAN1_LOW
        )
        self.disable()
        return full, ir

    def get_luminosity(self, channel):
        full, ir = self.get_full_luminosity()
        if channel == FULLSPECTRUM:
            # Reads two byte value from channel 0 (visible + infrared)
            return full
        elif channel == INFRARED:
            # Reads two byte value from channel 1 (infrared)
            return ir
        elif channel == VISIBLE:
            # Reads all and subtracts out ir to give just the visible!
            return full - ir
        else:  # unknown channel!
            return 0

    def get_current(self, format=''):
        full, ir = self.get_full_luminosity()
        lux = self.calculate_lux(full, ir)  # convert raw values to lux
        output = {
            'lux': lux,
            'full': full,
            'ir': ir,
            'gain': self.get_gain(),
            'integration_time': self.get_timing()
        }
        if format == 'json':
            return json.dumps(output)
        return output

    def test(self, int_time=INTEGRATIONTIME_100MS, gain=GAIN_LOW):
        self.set_gain(gain)
        self.set_timing(int_time)
        full_test, ir_test = self.get_full_luminosity()
        lux_test = self.calculate_lux(full_test, ir_test)
        print('Lux = {0:f}  full = {1}  ir = {2}'.format(lux_test, full_test, ir_test))
        print('Integration time = {}'.format(self.get_timing()))
        print('Gain = {} \n'.format(self.get_gain()))


tsl2591/tsl2591.py
'''
This code is an adaptation of the Arduino_TSL2591 library from
adafruit: https://github.com/adafruit/Adafruit_TSL2591_Library

For configuring I2C on Raspberry Pi
https://learn.adafruit.com/adafruits-raspberry-pi-lesson-4-gpio-setup/configuring-i2c

Datasheet
https://learn.adafruit.com/adafruit-tsl2591/downloads
'''
from __future__ import print_function

import time
import json
from smbus2 import SMBus

# *************************************************
# ******* MACHINE VARIABLES (DO NOT TOUCH) ********
# *************************************************
VISIBLE = 2  # channel 0 - channel 1
INFRARED = 1  # channel 1
FULLSPECTRUM = 0  # channel 0

ADDR = 0x29
READBIT = 0x01
COMMAND_BIT = 0xA0  # bits 7 and 5 for 'command normal'
CLEAR_BIT = 0x40  # Clears any pending interrupt (write 1 to clear)
WORD_BIT = 0x20  # 1 = read/write word (rather than byte)
BLOCK_BIT = 0x10  # 1 = using block read/write
ENABLE_POWERON = 0x01
ENABLE_POWEROFF = 0x00
ENABLE_AEN = 0x02
ENABLE_AIEN = 0x10
CONTROL_RESET = 0x80
LUX_DF = 408.0
LUX_COEFB = 1.64  # CH0 coefficient
LUX_COEFC = 0.59  # CH1 coefficient A
LUX_COEFD = 0.86  # CH2 coefficient B

REGISTER_ENABLE = 0x00
REGISTER_CONTROL = 0x01
REGISTER_THRESHHOLDL_LOW = 0x02
REGISTER_THRESHHOLDL_HIGH = 0x03
REGISTER_THRESHHOLDH_LOW = 0x04
REGISTER_THRESHHOLDH_HIGH = 0x05
REGISTER_INTERRUPT = 0x06
REGISTER_CRC = 0x08
REGISTER_ID = 0x0A
REGISTER_CHAN0_LOW = 0x14
REGISTER_CHAN0_HIGH = 0x15
REGISTER_CHAN1_LOW = 0x16
REGISTER_CHAN1_HIGH = 0x17
# *****************************************
# ******* END OF MACHINE VARIABLES ********
# *****************************************

# Integration time
# The integration time can be set between 100 and 600ms,
# and the longer the integration time the more light the
# sensor is able to integrate, making it more sensitive in
# low light the longer the integration time.
INTEGRATIONTIME_100MS = 0x00
INTEGRATIONTIME_200MS = 0x01
INTEGRATIONTIME_300MS = 0x02
INTEGRATIONTIME_400MS = 0x03
INTEGRATIONTIME_500MS = 0x04
INTEGRATIONTIME_600MS = 0x05

# Gain
# The gain can be set to one of the following values
# (though the last value, MAX, has limited use in the
# real world given the extreme amount of gain applied):
# GAIN_LOW: Sets the gain to 1x (bright light)
# GAIN_MEDIUM: Sets the gain to 25x (general purpose)
# GAIN_HIGH: Sets the gain to 428x (low light)
# GAIN_MAX: Sets the gain to 9876x (extremely low light)
GAIN_LOW = 0x00
GAIN_MED = 0x10
GAIN_HIGH = 0x20
GAIN_MAX = 0x30


class Tsl2591(object):
    '''
    An object class containing a series of methods to enable easy
    interaction with the sensor.
    '''

    def __init__(
            self,
            i2c_bus=1,
            sensor_address=ADDR,
            integration=INTEGRATIONTIME_200MS,
            gain=GAIN_MED
    ):
        self.bus = SMBus(i2c_bus)
        self.sender_address = sensor_address
        self.integration_time = integration
        self.gain = gain
        self.set_timing(self.integration_time)
        self.set_gain(self.gain)
        self.disable()  # to be sure

    def set_timing(self, integration):
        self.enable()
        self.integration_time = integration
        self.bus.write_byte_data(
            self.sender_address,
            COMMAND_BIT | REGISTER_CONTROL,
            self.integration_time | self.gain
        )
        self.disable()

    def get_timing(self):
        return self.integration_time

    def set_gain(self, gain):
        self.enable()
        self.gain = gain
        self.bus.write_byte_data(
            self.sender_address,
            COMMAND_BIT | REGISTER_CONTROL,
            self.integration_time | self.gain
        )
        self.disable()

    def get_gain(self):
        return self.gain

    def calculate_lux(self, full, ir):
        # Check for overflow conditions first
        if (full == 0xFFFF) | (ir == 0xFFFF):
            return 0

        case_integ = {
            INTEGRATIONTIME_100MS: 100.,
            INTEGRATIONTIME_200MS: 200.,
            INTEGRATIONTIME_300MS: 300.,
            INTEGRATIONTIME_400MS: 400.,
            INTEGRATIONTIME_500MS: 500.,
            INTEGRATIONTIME_600MS: 600.,
        }
        if self.integration_time in case_integ.keys():
            atime = case_integ[self.integration_time]
        else:
            atime = 100.

        case_gain = {
            GAIN_LOW: 1.,
            GAIN_MED: 25.,
            GAIN_HIGH: 428.,
            GAIN_MAX: 9876.,
        }

        if self.gain in case_gain.keys():
            again = case_gain[self.gain]
        else:
            again = 1.

        # cpl = (ATIME * AGAIN) / DF
        cpl = (atime * again) / LUX_DF
        lux1 = (full - (LUX_COEFB * ir)) / cpl

        lux2 = ((LUX_COEFC * full) - (LUX_COEFD * ir)) / cpl

        # The highest value is the approximate lux equivalent
        return max([lux1, lux2])

    def enable(self):
        self.bus.write_byte_data(
            self.sender_address,
            COMMAND_BIT | REGISTER_ENABLE,
            ENABLE_POWERON | ENABLE_AEN | ENABLE_AIEN
        )  # Enable

    def disable(self):
        self.bus.write_byte_data(
            self.sender_address,
            COMMAND_BIT | REGISTER_ENABLE,
            ENABLE_POWEROFF
        )

    def get_full_luminosity(self):
        self.enable()
        # not sure if we need it "// Wait x ms for ADC to complete"
        time.sleep(0.105+0.100*self.integration_time)
        full = self.bus.read_word_data(
            self.sender_address, COMMAND_BIT | REGISTER_CHAN0_LOW
        )
        ir = self.bus.read_word_data(
            self.sender_address, COMMAND_BIT | REGISTER_CHAN1_LOW
        )
        self.disable()
        return full, ir

    def get_luminosity(self, channel):
        full, ir = self.get_full_luminosity()
        if channel == FULLSPECTRUM:
            # Reads two byte value from channel 0 (visible + infrared)
            return full
        elif channel == INFRARED:
            # Reads two byte value from channel 1 (infrared)
            return ir
        elif channel == VISIBLE:
            # Reads all and subtracts out ir to give just the visible!
            return full - ir
        else:  # unknown channel!
            return 0

    def get_current(self, format=''):
        full, ir = self.get_full_luminosity()
        lux = self.calculate_lux(full, ir)  # convert raw values to lux
        output = {
            'lux': lux,
            'full': full,
            'ir': ir,
            'gain': self.get_gain(),
            'integration_time': self.get_timing()
        }
        if format == 'json':
            return json.dumps(output)
        return output

    def test(self, int_time=INTEGRATIONTIME_100MS, gain=GAIN_LOW):
        self.set_gain(gain)
        self.set_timing(int_time)
        full_test, ir_test = self.get_full_luminosity()
        lux_test = self.calculate_lux(full_test, ir_test)
        print('Lux = {0:f}  full = {1}  ir = {2}'.format(
            lux_test, full_test, ir_test))
        print('Integration time = {}'.format(self.get_timing()))
        print('Gain = {} \n'.format(self.get_gain()))
